Orange/projection/freeviz.py

```diff
@@ -0,0 +1,467 @@
+
+import numpy as np
+import scipy.spatial
+
+from Orange.data import ContinuousVariable, Domain
+from Orange.data.util import SharedComputeValue
+from Orange.projection import Projector, Projection
+
+
+class FreeVizProjector(SharedComputeValue):
+    """Transform into a given FreeViz component."""
+
+    def __init__(self, projection, feature, freeviz_transform):
+        super().__init__(freeviz_transform)
+        self.projection = projection
+        self.feature = feature
+        self.transformed = None
+
+    def compute(self, data, freeviz_space):
+        if self.feature >= freeviz_space.shape[1]:
+            return np.zeros((freeviz_space.shape[0]))
+        return freeviz_space[:, self.feature]
+
+
+class _FreeVizTransformDomain:
+    """Computation common for all FreeViz variables."""
+
+    def __init__(self, freeviz):
+        self.freeviz = freeviz
+
+    def __call__(self, data):
+        if data.domain != self.freeviz.pre_domain:
+            data = data.transform(self.freeviz.pre_domain)
+        return self.freeviz.transform(data.X)
+
+
+class _LinearCombination:
+    def __init__(self, attrs, weights, mean=None):
+        self.attrs = attrs
+        self.weights = weights
+        self.mean = mean
+
+
+class FreeVizModel(Projection):
+    name = "FreeVizModel"
+
+    def __init__(self, proj, domain):
+        freeviz_transform = _FreeVizTransformDomain(self)
+
+        def freeviz_variable(i):
+            v = ContinuousVariable(
+                "FreeViz Component {}".format(i + 1),
+                compute_value=FreeVizProjector(self, i, freeviz_transform))
+            v.to_sql = _LinearCombination(
+                domain.attributes, self.components_[i, :],
+                getattr(self, 'mean_', None))
+            return v
+
+        super().__init__(proj=proj)
+        self.orig_domain = domain
+        self.n_components = self.components_.shape[0]
+        self.domain = Domain(
+            [freeviz_variable(i) for i in range(self.n_components)],
+            domain.class_vars, domain.metas)
+
+
+class FreeViz(Projector):
+    name = 'FreeViz'
+    supports_sparse = False
+
+    def __init__(self, weights=None, center=True, scale=True, dim=2, p=1,
+                 initial=None, maxiter=500, alpha=0.1, atol=1e-5, preprocessors=None):
+        super().__init__(preprocessors=preprocessors)
+        self.params = vars()
+        self.weights = weights
+        self.center = center
+        self.scale = scale
+        self.dim = dim
+        self.p = p
+        self.initial = initial
+        self.maxiter = maxiter
+        self.alpha = alpha
+        self.atol = atol
+        self.is_class_discrete = False
+
+    def __call__(self, data):
+        if data and data.domain.class_var.is_discrete:
+            self.is_class_discrete = True
```

Comment:
The same instance of `FreeViz` can be called multiple times and with different data sets, hence it can't count that `is_class_discrete` is `False` (as set in the constructor`.

```
if data is not None:
    self.is_class_discrete = data.domain.class_var.is_discrete
```

Reply:
Fixed.